src/tokens/ERC5095.sol

```diff
@@ -196,17 +220,95 @@ contract ERC5095 is ERC20Permit, IERC5095 {
+        // Pass the received shares onto the intended receiver
+        _transfer(address(this), r, returned);
+
+        return returned;
+    }
+
+    /// @notice Before maturity spends `assets` of underlying, and sends `shares` of PTs to `receiver`. Post or at maturity, reverts.
+    /// @param r The receiver of the principal tokens
+    /// @param a The amount of underlying tokens deposited
+    /// @return uint256 The amount of principal tokens burnt by the withdrawal
+    function deposit(address r, uint256 a) external override returns (uint256) {
+        // Revert if called at or after maturity
+        if (block.timestamp >= maturity) {
+            revert Exception(
+                21,
+                block.timestamp,
+                maturity,
+                address(0),
+                address(0)
+            );
         }
 
+        // Preview how many shares are needed to withdraw the desired amount of underlying
+        uint128 shares = Cast.u128(previewDeposit(a));
+
+        // Receive the funds from the sender
+        Safe.transferFrom(IERC20(underlying), msg.sender, address(this), a);
+
+        // consider the hardcoded slippage limit, 4626 compliance requires no minimum param.
+        uint128 returned = IMarketPlace(marketplace).sellUnderlying(
+            underlying,
+            maturity,
+            Cast.u128(a),
+            shares
+        );
+
         // Pass the received shares onto the intended receiver
         _transfer(address(this), r, returned);
 
         return returned;
     }
 
+    /// @notice Before maturity mints `shares` of PTs to `receiver` by spending underlying. Post or at maturity, reverts.
+    /// @param r The receiver of the underlying tokens being withdrawn
+    /// @param s The amount of underlying tokens withdrawn
+    /// @param m Minimum amount of shares that the user must receive
+    /// @return uint256 The amount of principal tokens burnt by the withdrawal
+    function mint(address r, uint256 s, uint256 m) external returns (uint256) {
+        // Revert if called at or after maturity
+        if (block.timestamp >= maturity) {
+            revert Exception(
+                21,
+                block.timestamp,
+                maturity,
+                address(0),
+                address(0)
+            );
+        }
+
+        // Calculate how much underlying will be needed to mint the desired shares
+        uint128 assets = Cast.u128(previewMint(s));
+
+        // Confirm that slippage has not been exceeded
+        if (m > assets) {
+            revert Exception(16, m, assets, address(0), address(0));
+        }
+
+        // Transfer the underlying to the token
+        Safe.transferFrom(
+            IERC20(underlying),
+            msg.sender,
+            address(this),
+            assets
+        );
+
+        // Swap the underlying for principal tokens via the pool
+        uint128 returned = IMarketPlace(marketplace).sellUnderlying(
```

Comment:
Maybe we use `buyPrincipalToken` for mint because the params were provided match it? 
https://github.com/sherlock-audit/2022-10-illuminate/blob/main/src/Marketplace.sol#L323

With that the case, I think we just do:

```
uint128 returned = IMarketPlace(marketplace).buyPrincipalToken(
            underlying,
            maturity,
            s,
            m)
```

This allows us to actually remove the slippage check on 285, and rely on the slippage controls in marketplace?